modyn/selector/internal/selector_strategies/presampling_strategies/label_balanced_presampling_strategy.py

```diff
@@ -0,0 +1,141 @@
+from typing import Optional
+
+from modyn.metadata_database.metadata_database_connection import MetadataDatabaseConnection
+from modyn.metadata_database.models import SelectorStateMetadata
+from modyn.selector.internal.selector_strategies.presampling_strategies import AbstractPresamplingStrategy
+from sqlalchemy import Select, asc, func, select
+from sqlalchemy.sql.expression import func as sql_func
+
+
+def get_fair_share(capacity: int, requests: list[int]) -> int:
+    """
+    Simple Fair Sharing algorithm.
+    Here it's used to determine how many samples gets each class (min(fair_share, num_samples)
+    """
+    fair_share = int(capacity / len(requests))
+    below_fair_share = [requests.index(el) for el in requests if el <= fair_share]
+
+    if len(below_fair_share) != 0 and len(requests) != len(below_fair_share):
+        new_capacity = capacity - sum(requests[i] for i in below_fair_share)
+        new_requests = [requests[i] for i in range(len(requests)) if i not in below_fair_share]
+        return get_fair_share(new_capacity, new_requests)
+    return fair_share
+
+
+def get_fair_share_predicted_total(fair_share: int, requests: list[int]) -> int:
+    return sum(min(fair_share, req) for req in requests)
+
+
+class LabelBalancedPresamplingStrategy(AbstractPresamplingStrategy):
```

Comment:
this (and the other strategies) are cool, but we should also test they work in postgres. could you add some meaningful integrationtests (there are already examples for selector integration tests) to all new presampling strategies, showing they work somewhat as inteded? (i know its hard to test randomness but we can do assertions for the stuff we can test)

Reply:
please also leave this comment open for me as a reminder to go through this class again. I am very tired right now and cannot really check the queries too detailed